src/lib.rs

```diff
@@ -473,29 +475,48 @@ impl<L: Ledger> Default for EventSummary<L> {
 }
 
 impl<'a, L: 'static + Ledger> KeystoreState<'a, L> {
-    pub fn balance(
+    pub fn balance_for_key<Meta: Serialize + DeserializeOwned + Send>(
         &self,
+        model: &KeystoreModel<'a, L, impl KeystoreBackend<'a, L>, Meta>,
+        asset: &AssetCode,
+        address: &UserAddress,
+        frozen: FreezeFlag,
+    ) -> U256 {
+        let spendable = model.records.get_spendable::<L>(asset, address, frozen);
+        if let Some(records) = spendable {
+            records
+                .filter(move |record| !record.on_hold(self.txn_state.block_height()))
+                .fold(U256::zero(), |sum, record| sum + record.amount())
+        } else {
+            U256::zero()
+        }
+    }
+
+    pub fn balance<Meta: Serialize + DeserializeOwned + Send>(
+        &self,
+        model: &KeystoreModel<'a, L, impl KeystoreBackend<'a, L>, Meta>,
         addresses: impl Iterator<Item = UserAddress>,
         asset: &AssetCode,
         frozen: FreezeFlag,
     ) -> U256 {
         let mut balance = U256::zero();
         for address in addresses {
-            balance += self.txn_state.balance(asset, &address, frozen);
+            balance += self.balance_for_key(model, asset, &address, frozen);
         }
         balance
     }
 
-    pub fn balance_breakdown(
+    pub fn balance_breakdown<Meta: Serialize + DeserializeOwned + Send>(
```

Comment:
Looks like this function is the same as `balance_for_key`. Can we keep only one? It's probably simpler to keep `balance_breakdown` since it's called in many places.

Reply:
Yes good catch, this got mixed up during merging, I'll keep only `balance_breakdown`